Add StripeMetadataSourceFactory to consume loaded DWRF stripe cache

sdruzkin opened the pull request:
This change allows OrcReader to populate DWRF stripe cache using the stripe cache slices from the OrcFileTail and the CacheOffsets from the footer. The created DWRF stripe cache then passed into the StripeMetadataSource via a new interface StripeMetadataSourceFactory. This new interface is added after the discussion with @shixuan-fan and @highker  to avoid adding mutating methods into StripeMetadataSource, and still be able to pass DWRF stripe cache directly from the file/source into the StripeMetadataSource.

At this point the DWRF stripe cache can be already used if we add a new implementation of StripeMetadataSource that uses the DWRF stripe cache and falls back to the StorageStripeMetadataSource. I plan to add this new class in a next PR to keep the scope of this PR manageable.

This is a continuation of https://github.com/prestodb/presto/pull/16114

Test plan:
- add new unit test that reads files created by BBIO with various DWRF stripe cache settings

```
== NO RELEASE NOTE ==
```

sdruzkin commented on presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java:
```diff
@@ -1458,7 +1458,6 @@ static OrcBatchRecordReader createCustomOrcRecordReader(
                 orcEncoding,
                 orcFileTailSource,
                 stripeMetadataSource,
-                Optional.empty(),
```
Due to some unknown reason this was using the package protected method. I decided to move it to a public method instead of wrapping the stripeMetadataSource into the StripeMetadataSourceFactory.of(stripeMetadataSource)

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -94,7 +100,32 @@ public OrcReader(
                 orcEncoding,
                 orcFileTailSource,
-                stripeMetadataSource,
+                StripeMetadataSourceFactory.of(stripeMetadataSource),
```
I kept the old constructor and added a new constructor with the StripeMetadataSourceFactory. I kept the old constructor to avoid changing multiple (10+) files not really related to this change. Depending on the reviewer's feedback I can keep it as is if it's not a big deal, or I cab make changes in all other places in a separate house-cleaning PR.

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -186,6 +215,59 @@ public OrcReader(
+    private DwrfStripeCache buildDwrfStripeCache(List<StripeInformation> stripes, DwrfStripeCacheData dwrfStripeCacheData, List<Integer> cacheOffsets)
+    {
+        List<StripeInformation> sortedStripes = new ArrayList<>(stripes);
+        sortedStripes.sort(comparingLong(StripeInformation::getOffset));
```
This is a bit controversial, but it's coming from https://github.com/prestodb/presto/blob/master/presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java#L194-L203. I don't know what is the real reason for sorting, it's a bit hard to do blame on this file because it was refactored a couple times.

We don't need sorting in FB use cases, I'm open to nuke sorting from this place.

arunthirupathi replied:
Can we please add a comment to capture this reason ?

sdruzkin replied:
Done, it's not very convincing thou :)

sdruzkin commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -186,6 +215,59 @@ public OrcReader(
+
+    private static boolean addDwrfStripeCacheSlice(int cacheOffsetIdx, List<Integer> cacheOffsets, Slice cacheSlice, ImmutableMap.Builder<Long, Slice> slices, StripeInformation stripe)
+    {
+        if (cacheOffsetIdx >= cacheOffsets.size() + 1) {
```
You are right -fixed. I'll add a nice unit test to this PR by noon.

sdruzkin commented:
Messed up commit history again. Working on fixing it.

sdruzkin commented:
Addressed all comments. Working on porting a unit test from the grandfather PR.

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -186,6 +215,61 @@ public OrcReader(
+
+    private static boolean addDwrfStripeCacheSlice(int cacheOffsetIdx, List<Integer> cacheOffsets, Slice cacheSlice, ImmutableMap.Builder<Long, Slice> slices, StripeInformation stripe)
+    {
+        if (cacheOffsetIdx >= cacheOffsets.size() - 1) {
```
I would recommend this to be changed to cacheOffsetIdx + 1 < cacheOffsets.size() make it consistent, down below we are accessing + 1 than the idx passed in.

sdruzkin replied:
done

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -186,6 +215,61 @@ public OrcReader(
+
+        int sliceOffset = cacheOffsets.get(cacheOffsetIdx);
+        int sliceSize = cacheOffsets.get(cacheOffsetIdx + 1) - sliceOffset;
+        checkState(sliceOffset < cacheSlice.length());
```
add, some debug information on what failed the checkState.

sdruzkin replied:
done

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheMode.java:
```diff
@@ -19,14 +19,65 @@
+     * Get the position of an element in the cacheOffsets that can be used to
+     * get the offset for the index streams slice for a given stripe.
+     */
+    public int getIndexOffsetPosition(int stripeOrdinal)
```
Can we add unit test for getIndexoffsetPostiion and getFooterOffsetPosition ?

sdruzkin replied:
Done via a new comprehensive test

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCache.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Get a slice containing an index slice for a given stripe.
+     */
+    public Optional<Slice> getIndexStreamsSlice(StripeReader.StripeId stripeId)
```
use static import for StripeId

sdruzkin replied:
done

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCache.java:
```diff
@@ -0,0 +1,84 @@
+    public DwrfStripeCache(DwrfStripeCacheMode mode, Map<Long, Slice> indexSlices, Map<Long, Slice> footerSlices)
+    {
+        this.mode = mode;
+        this.indexSlices = indexSlices;
```
verify the parameters are not null.

sdruzkin replied:
done

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheMode.java:
```diff
@@ -19,14 +19,65 @@
+    private final boolean hasIndex;
+    private final boolean hasFooter;
+
+    DwrfStripeCacheMode(boolean hasIndex, boolean hasFooter)
```
should this be private constructor ?

arunthirupathi replied:
If no access modifier is specified enum defaults to private constructor, so no changes is required here. https://stackoverflow.com/a/7748220

sdruzkin replied:
You are right. Enums are sealed, so making it private is redundant.

arunthirupathi commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheMode.java:
```diff
@@ -19,14 +19,65 @@
+            case INDEX_AND_FOOTER:
+                return stripeOrdinal * 2;
+            default:
+                throw new IllegalStateException("This mode doesn't support index cache: " + this);
```
This class does not have toString implementation, but I assume enum has default toString implementation, just verifying that is the case here.

sdruzkin replied:
Correct, by default enum.toString looks exactly the same as enum.name()

arunthirupathi reviewed (commented):
Took a detailed look and left some comments. The PR looks good once addressed.

sdruzkin commented:
Added a new test with a bunch of files created using bbio. Addressed all comments from the recent batch.

arunthirupathi reviewed (approved):
Thanks for adding the comprehensive unit tests, this looks good to me.

bhhari commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -186,6 +215,67 @@ public OrcReader(
+
+    private static boolean addDwrfStripeCacheSlice(int cacheOffsetIdx, List<Integer> cacheOffsets, Slice cacheSlice, ImmutableMap.Builder<Long, Slice> slices, StripeInformation stripe)
+    {
+        if (!(cacheOffsetIdx + 1 < cacheOffsets.size())) {
```
instead of negation can we just use
cacheOffsetIdx >= cacheOffsets.size()

arunthirupathi replied:
The original condition was, cacheoffsetIdx >= cacheOffsets.size() -1, I asked to modify the condition as I found the code below is using cacheoffsetIdx + 1 to access the element and using the same in the if loop would make it easy for the reader ( at least for me, as they are the same condition).
```
boolean isValidIndex = (cacheOffsetIdx + 1) < cacheOffsets.size();
if (!isValidIndex) {
    return false;
}
```

sdruzkin replied:
Like the isValidIndex approach, updated PR with this change.

bhhari reviewed (approved):
LGTM

shixuan-fan commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java:
```diff
@@ -186,6 +215,68 @@ public OrcReader(
+        if (mode.hasIndex()) {
+            for (int i = 0; i < sortedStripes.size(); i++) {
+                int offsetIdx = mode.getIndexOffsetPosition(i);
+                if (!addDwrfStripeCacheSlice(offsetIdx, cacheOffsets, cacheSlice, indexSlices, sortedStripes.get(i))) {
```
Can we extract the valid check from `addDwrfStripeCacheSlice` to a separate method `isValidOffset`? This way I feel is a little bit cleaner that we don't need to return `boolean` for an adding action, and the logic in this for loop would become:

```
for (...) {
  if (!isValidOffset(offset)) {
    break;
  }
  addDwrfStripeCacheSlice(...);
}
```

sdruzkin replied:
done

shixuan-fan commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCache.java:
```diff
@@ -0,0 +1,85 @@
+        if (mode.hasIndex()) {
+            Slice slice = indexSlices.get(stripeId.getOffset());
+            if (slice != null) {
+                return Optional.of(slice);
```
nit: `return Optional.ofNullable(slice)` so we don't need the non-null check.

sdruzkin replied:
done

shixuan-fan commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheMode.java:
```diff
@@ -19,14 +19,65 @@
+            case INDEX:
+                return stripeOrdinal;
+            case INDEX_AND_FOOTER:
+                return stripeOrdinal * 2;
```
Dumb question: why is it `stripeOrdinal * 2` here and `stripeOrdinal * 2 + 1` below for footer offset?

sdruzkin replied:
If the mode is set to BOTH then offsets would be saved in pairs for each stripe: 
```
Stripe 0 index offset
Stripe 0 footer offset
...
Stripe 0 index offset   <-- stripe idx * 2 + 0
Stripe 0 footer offset  <-- stripe idx * 2 + 1
```

shixuan-fan reviewed (approved):
LGTM. Mostly nits :)

sdruzkin commented:
I moved construction of DwrfStripeCache out of OrcReader into the  DwrfStripeCacheData, after all the utility methods we added it seems like a more natural place for it.

sdruzkin commented:
Rebase

sdruzkin commented:
ci / test (:presto-tests -P presto-tests-execution-memory) failed with:
```
Error:  Tests run: 35, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 422.515 s <<< FAILURE! - in TestSuite
Error:  testExceedSoftLimits(com.facebook.presto.execution.TestQueues)  Time elapsed: 11.11 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method com.facebook.presto.execution.TestQueues.testExceedSoftLimits() didn't finish within the time-out 240000
```

It doesn't seem to be related.

sdruzkin commented:
@rschlussel can you please take a look at this PR when you have time?
